### PR TITLE
[CQT-291] Implement `BasisVector` as a `boost::dynamic_bitset<uint32_t>`

### PR DESCRIPTION
**Pull request by rturrado**
Why not to use the same underlying type for all of our bitsets, i.e., `BasisVector` and the measurement registers?
We may (arguably) lose a little bit in performance terms.
But we gain in consistency and simplicity.
Apart from the fact that we are not limited to working always with a `BasisVector` of `MAX_QUBIT_NUMBER`.

Implement `BasisVector` as a `boost::dynamic_bitset<uint32_t>`:
- Remove custom `Bitset` class.
- Remove `bitset_test.cpp`.

Add `MeasurementRegister` alias (yet another boost dynamic bitset).

### CODE REVIEW

File: include/qx/core.hpp [hunk 1]
```diff
@@ -32,13 +31,12 @@ struct BitIndex {
     std::size_t value;
 };
 
-using BasisVector = utils::Bitset<config::MAX_QUBIT_NUMBER>;
-
+using BasisVector = boost::dynamic_bitset<uint32_t>;
+using MeasurementRegister = boost::dynamic_bitset<uint32_t>;
 using BitMeasurementRegister = boost::dynamic_bitset<uint32_t>;
```
Comment (elenbaasc) [inline include/qx/core.hpp:36]: Just for my understanding: Should we rename the BitMeasurementRegister, simply the BitRegister? (Maybe not part of this PR)

**Review by elenbaasc — APPROVED** [review]
Looks all good!

I cannot attest to the decrease in performance, so I presume that it shouldn't be too significant.

Just one question from my side.

File: include/qx/core.hpp (hunk 1 above)
Comment (rturrado) [reply include/qx/core.hpp:36]: I wouldn't say so.

On one hand, the RegisterManager  holds a qubit register and a **bit register**. Those are data structures that keep the mappings between variables and a virtual qubit or bit register (i.e., index _b<sub>i</sub>_ of bit register variable `b` is in position _vb<sub>j</sub>_ of virtual bit register).

On the other hand, we manage **measurement registers**. In the future, we should only manage the bit measurement register, which could then be renamed to measurement register. But, at the moment, because there is no agreement yet on the bit measurement register, we are keeping the _legacy_ measurement register. This one assumes that every qubit in the virtual qubit register is measured to the same position in the measurement register.

For example:
- index _q<sub>i</sub>_ of qubit register variable `q` is in position _vq<sub>j</sub>_ of virtual qubit register.
- You measure it to index _b<sub>m</sub>_ of bit register variable `b`, which is in position _vb<sub>n</sub>_ of virtual bit register.
- You'll update position _mr<sub>j</sub>_ of measurement register, but position _bmr<sub>n</sub>_ of bit measurement register.

I.e., when using the measurement register we behave as if there were no bit variables, just qubit variables and a virtual qubit register.
